File: 05-NumericalIntegration/2-trap-richardson.py
"""
Description: Numerical Integration methods: Richardson-Trapezoidal Rule
"""
import numpy as np
import matplotlib.pyplot as plt

#Function to integrate
def f(x):
    return np.sin(x)
    
#Trapezoidal rule: integral from x0 to xf with n steps
def trapezoidal(fun, x0, xf, n):
    h = (xf - x0)/n
    suma = 0.5*(fun(x0) + fun(xf))
    for i in np.arange(n):
        xi = x0 + i*h
        suma += fun(xi)
    suma *= h
    return suma

#Richardson rule to improve trapezoidal rule
def richardson(alg, fun, x0, xf, n):
    return (4*alg(fun, x0, xf, 2*n) -alg(fun, x0, xf, n))/3

#Parameters and printing of the results
Xmin = 0.0
Xmax = 1.0

for i in np.arange(7):
    Nmax = pow(10, i)
    numerical_trap = trapezoidal(f, Xmin, Xmax, Nmax)
    numerical_rich = richardson(trapezoidal, f, Xmin, Xmax, Nmax)
    theoretical = -np.cos(Xmax) + np.cos(Xmin)
    Err_trap = abs(1 - numerical_trap/theoretical)
    Err_rich = abs(1 - numerical_rich/theoretical)
    print(Nmax, Err_trap, Err_rich)







"""
//////////////////////////////////////////////////////////////////////////
Created on Fri Jan  1 13:32:11 2021                                    //
                                                                      //
@author: Nicolás Gómez                                               //
//////////////////////////////////////////////////////////////////////
"""

File: 05-NumericalIntegration/2-trap-richardson.cpp
#include <iostream>
#include <cmath>
#include <functional>

using fptr  = double(double);
using fptr2 = double(fptr, double, double, int);

double f(double x);
double trapecio(fptr fun, double a, double b, int N);
double richardson(fptr2 alg, fptr fun, double a, double b, int N);

int main(int argc, char **argv)
{
    std::cout.precision(15); std::cout.setf(std::ios::scientific);
    
    const double xmin = 0.0;
    const double xmax = 1.0;

    for(int NMAX = 1; NMAX <= 1.0e8; NMAX *= 10) {
        double intnum_trapecio = trapecio(f, xmin, xmax, NMAX);
        double intnum_traprich = richardson(trapecio, f, xmin, xmax, NMAX);
        double inttheo = -std::cos(xmax) + std::cos(xmin);
        double delta_trapecio = std::fabs(1 - intnum_trapecio/inttheo);
        double delta_traprich = std::fabs(1 - intnum_traprich/inttheo);
        std::cout << NMAX
                  << "\t" << delta_trapecio
                  << "\t" << delta_traprich
                  << "\n";
    }
    
    return 0;
}

double f(double x)
{
    return std::sin(x);
}

double trapecio(fptr fun, double a, double b, int N)
{
    const double h = (b-a)/N;
    double suma = 0.5*(fun(a) + fun(b));
    for(int ii = 1; ii <= N-1; ++ii) {
        double xi = a + ii*h;
        suma += fun(xi);
    }
    suma *= h;
    return suma;
}

double richardson(fptr2 alg, fptr fun, double a, double b, int N)
{
    return (4*alg(fun, a, b, 2*N) - alg(fun, a, b, N))/3;
}
